clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -2036,8 +2271,30 @@ public void resume(Collection<TopicPartition> partitions) {
      */
     @Override
     public Map<TopicPartition, Long> endOffsets(Collection<TopicPartition> partitions, Duration timeout) {
+        return endOffsets(partitions, timeout, false);
+    }
+
+    protected Map<TopicPartition, Long> endOffsets(Collection<TopicPartition> partitions, Duration timeout, final boolean isChildConsumer) {
         acquireAndEnsureOpen();
         try {
+            if (!isChildConsumer && useParallelRebalance) {
+                checkRebalance();
+
+                rebalanceConsumer.sendRequest(timeout, RebalanceKafkaConsumer.ConsumerRequest.END_OFFSETS, partitions, new DefaultTaskCompletionCallback());
+                final Map<TopicPartition, Long> offsets1 = fetcher.endOffsets(partitions, timeout.toMillis());
+                final Map<TopicPartition, Long> offsets2 = (Map<TopicPartition, Long>) pollForResults(timeout.toMillis(), time.milliseconds()).value;
+                final Map<TopicPartition, Long> result = new HashMap<>();
+                for (final TopicPartition partition : partitions) {
```

Comment:
See if you can refactor this code which is similar to what beginningOffsets has (apart from the condition between pos1 and pos2)